### PR TITLE
bpo-35346, platform: replace os.popen() with subprocess

### PR DESCRIPTION
**Pull request by vstinner**
Replace os.popen() with subprocess.Popen in the platform module:

* Pass the command as a list of strings, not as as a string;
* Don't use a shell;
* _syscmd_ver() now redirects errors to DEVNULL.

https://bugs.python.org/issue35346

### CODE REVIEW

**Review by serhiy-storchaka — COMMENTED** [review]
Can `subprocess.check_output()` be used here?

**Comment by vstinner** [comment]
> Can subprocess.check_output() be used here?

This function is documented in the "Older high-level API" section of the subprocess documentation. It seems like subprocess.run() is now preferred. subprocess.run() kills the process on exception, IMHO it's a nice enhancement (to not leak a process on error). I modified my PR to use subprocess.run().

**Comment by malemburg** [comment]
+1 on using .check_output() instead. It's a special purpose API for exactly this use case and also available in Python 2.7.

**Comment by serhiy-storchaka** [comment]
I think the code would be simpler, and therefore more readable, if use `check_output()`.

**Review by serhiy-storchaka — COMMENTED** [review]
Visible effect: suppressed the output to stderr. I think it is worth a news entry. And it can be worth backporting.

**Comment by vstinner** [comment]
> I think the code would be simpler, and therefore more readable, if use check_output().

done

> Visible effect: suppressed the output to stderr. I think it is worth a news entry.

done

> And it can be worth backporting.

My PR only changes _syscmd_ver() when os.uname() doesn't exist, so only impacts Windows. I don't recall that I ever saw any spurious error when using platform on Windows.

I would prefer to minimize changes in stable branches. So if someone considers that it's worth it (I don't think so), I suggest to add " 2> %s" % DEV_NULL to _syscmd_ver() command, rather than using check_output().

**Comment by vstinner** [comment]
Hum, my PR didn't work as expected on Windows. Currently, _syscmd_ver() tries to run "ver" (as a string) with os.open() which uses a shell. My PR avoids shell=True. _syscmd_ver() still worked thanks to "cmd /c ver"... but that's basically like calling "ver" with shell=True, since subprocess uses cmd.exe as the shell:

```
            if shell:
                startupinfo.dwFlags |= _winapi.STARTF_USESHOWWINDOW
                startupinfo.wShowWindow = _winapi.SW_HIDE
                comspec = os.environ.get("COMSPEC", "cmd.exe")
                args = '{} /c "{}"'.format (comspec, args)
```

subprocess may not use cmd.exe if COMSPEC is set, but cmd.exe should always be available on Windows, no?

File: Lib/platform.py
Comment (serhiy-storchaka) [inline Lib/platform.py:617]: The code contradicts the docstring. This can lead to returning an incorrect result if the executable path contains some key strings like "'32-bit'" or "PE". This is a separate issue, but it could be better to fix it before merging this PR. Merging this PR will make backporting harder.

Comment (vstinner) [reply Lib/platform.py:617]: It has been changed by commit 685fffa8f427b3a768b232170752565d58c32112, https://bugs.python.org/issue16112 The docstring seems to be outdated.

Comment (vstinner) [reply Lib/platform.py:617]: > This can lead to returning an incorrect result if the executable path contains some key strings like "'32-bit'" or "PE".

I would prefer to keep this PR simple, and so don't try to fix this bug nor update the docstring.

Comment (vstinner) [reply Lib/platform.py:617]: @serhiy-storchaka created https://bugs.python.org/issue35348

File: Lib/platform.py
Comment (serhiy-storchaka) [inline Lib/platform.py]: Why an AttributeError is caught here?

**Comment by serhiy-storchaka** [comment]
Thank you, it looks better. But please defer merging it until issues with `_syscmd_file()` be resolved.

File: Lib/platform.py
Comment (vstinner) [reply Lib/platform.py]: AttributeError is to catch... missing os.popen() function (I guess) :-) AttributeError is as old as the module itself: commit 246d847475d041a823dbc3cc3aceae46e97de937. Do you think that we should remove it?

Comment (serhiy-storchaka) [reply Lib/platform.py]: I think it is better to remove it.

Comment (malemburg) [reply Lib/platform.py]: Guys, please remember that these modules are used by other Python implementations as well. The AttributeError most likely came from this: having e.g. IronPython or Jython not having the needed os API available.

Comment (vstinner) [reply Lib/platform.py]: Other Python implementations are supposed the implement the same stdlib. Why would IronPython or Jython not have subprocess.check_output()? For example, if IronPython implements "Python 3.5", they should copy platform.py from CPython 3.5.

If they decide to not implement subprocess.check_output(), well, we can have a talk :-) In the meanwhile, PyPy does patch the stdlib, but they don't propose these patches upstream :-(

Comment (malemburg) [reply Lib/platform.py]: This used to be different... could be that we didn't listen to them.

In any case, see here for when the AttributeErrors were added: https://github.com/python/cpython/blob/2b665dbb53d02be09947116c1fe2ac002375d348/Lib/platform.py#L64

In Python 2 the module was meant to work across many different Python versions and platforms. It would be good to keep this feature in place, since it needs to be updated more frequently than the whole Python version stack (i.e. applications should be able to copy a more recent from from 3.7 to 3.5 and still have it work).

Comment (vstinner) [reply Lib/platform.py]: Well, subprocess and and os.popen() are now available available on all platforms. The situation became better in Python 3 ;-)

Comment (vstinner) [reply Lib/platform.py]: Note: PyPy3 has os.popen() and subprocess, and Jython and IronPython are both stuck at Python 2.7 ;-)

Comment (serhiy-storchaka) [reply Lib/platform.py]: `os.popen()` is optional in Python 2, but always available in Python 3. If other implementation have enough different stdlib without `os.popen()` and `subprocess.check_output()`, it can have different `platform` module as well.

I believe it is safe to remove AttributeError in all Python 3 versions.

Comment (vstinner) [reply Lib/platform.py]: Ok. I removed the AttributeError. Anyway, if someone comes with an usecase where catching AttributeError makes sense, I have no problem to reintroduce it ;-) We can even revert this whole commit later if it causes any issue ;-)

File: Lib/platform.py
Comment (serhiy-storchaka) [inline Lib/platform.py]: Why not use just `['ver']` with `shell=True`?

Comment (vstinner) [reply Lib/platform.py]: > Why not use just ['ver'] with shell=True?

I wanted to avoid shell=True whenever possible, but ok, let's use shell=True here.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM if it LGT @malemburg.

**Comment by vstinner** [comment]
@malemburg: I plan to merge this PR at the end of the week.

**Comment by vstinner** [comment]
I squashed the 7 commits into a single commit to update the commit message, I rebased my PR and I also replaced os.open() with subprocess.check_output() in test_platform.

**Comment by pablogsal** [comment]
I removed the comments of the buildbot failures because they were false positives due to bad git checkouts. I am investigating why [this PR that we made to fix that problem](https://github.com/python/buildmaster-config/pull/69) did not work.

**Pull request merged**
